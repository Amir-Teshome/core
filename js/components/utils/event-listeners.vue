<script>
/**
 * Global event listeners (basically, those without a Vue instance access) go here.
 */
import isMobile from 'ismobilejs'
import router from '@/router'
import { ls } from '@/services'
import { playlistStore, preferenceStore, recentlyPlayedStore, userStore } from '@/stores'
import { alerts, event, forceReloadWindow } from '@/utils'

const deletePlaylist = playlist => {
  const destroy = async () => {
    await playlistStore.delete(playlist)
    alerts.success(`Deleted playlist &quot;${playlist.name}&quot;.`)
    router.go('home')
  }

  if (!playlist.songs.length) {
    destroy()
    return
  } else {
    alerts.confirm(`Delete the playlist &quot;${playlist.name}&quot?`, destroy)
  }
}

export default {
  render: h => null,

  created () {
    event.on({
      [event.$names.PLAYLIST_DELETE]: playlist => deletePlaylist(playlist),

      /**
       * Log the current user out and reset the application state.
       */
      async [event.$names.LOG_OUT] () {
        await userStore.logout()
        ls.remove('jwt-token')
        forceReloadWindow()
      },

      [event.$names.KOEL_READY]: () => router.init(),

      /**
       * Listen to 'main-content-view:load' event to load all recently played songs into the view
       */
      [event.$names.LOAD_MAIN_CONTENT]: async view => {
        if (view === 'recently-played') {
          recentlyPlayedStore.fetchAll()
        }
      },

      /**
       * Hide the panel away if a main view is triggered on mobile.
       */
      [event.$names.LOAD_MAIN_CONTENT]: () => (isMobile.phone && (preferenceStore.showExtraPanel = false))
    })
  }
}
</script>
