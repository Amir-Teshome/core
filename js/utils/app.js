// App (electron)-only methods
let mainWindow

if (KOEL_ENV === 'app') {
  mainWindow = require('electron').remote.getCurrentWindow()
}

export const app = {
  triggerMaximize: () => mainWindow && (mainWindow.isMaximized() ? mainWindow.unmaximize() : mainWindow.maximize())
}
