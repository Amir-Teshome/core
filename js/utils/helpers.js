export const use = (value, cb) => typeof value !== 'undefined' && value !== null && cb(value)
