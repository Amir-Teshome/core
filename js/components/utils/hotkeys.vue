<template>
  <global-events
    @keydown.space = "togglePlayback"
    @keydown.j = "playNext"
    @keydown.k = "playPrev"
    @keydown.f = "search"
    @keydown.l = "toggleLike"
    @keydown.mediaPrev = "playPrev"
    @keydown.mediaNext = "playNext"
    @keydown.mediaToggle = "togglePlayback"
  />
</template>

<script>
import Vue from 'vue'
import { $, event } from '@/utils'
import { playback, socket } from '@/services'
import { queueStore, favoriteStore, songStore } from '@/stores'

let ipc, events

if (KOEL_ENV === 'app') {
  ipc = require('electron').ipcRenderer
  events = require('&/events').default
}

// Register our custom key codes
Vue.config.keyCodes = {
  a: 65,
  j: 74,
  k: 75,
  f: 70,
  l: 76,
  mediaNext: 176,
  mediaPrev: 177,
  mediaToggle: 179
}

/**
 * Listen to the global shortcuts (media keys).
 * Only works in the app.
 */
let listenToGlobalShortcuts = () => {}

if (KOEL_ENV === 'app') {
  listenToGlobalShortcuts = () => {
    const mediaFunctionMap = {
      MediaNextTrack: () => playback.playNext(),
      MediaPreviousTrack: () => playback.playPrev(),
      MediaStop: () => playback.stop(),
      MediaPlayPause: () => playback.toggle()
    }

    ipc.on(events.GLOBAL_SHORTCUT, (e, msg) => msg in mediaFunctionMap && mediaFunctionMap[msg]())
  }
}

export default {
  methods: {
    /**
     * Toggle playback when user presses Space key.
     *
     * @param {Object} e The keydown event
     */
    togglePlayback: e => {
      if (e && $.is(e.target, 'input, textarea, button, select') && !$.is(e.target, 'input[type=range]')) {
        return true
      }

      playback.toggle()
      e && e.preventDefault()
    },

    /**
     * Play the previous song when user presses K.
     *
     * @param {Object} e The keydown event
     */
    playPrev: e => {
      if ($.is(e.target, 'input, select, textarea')) {
        return true
      }

      playback.playPrev()
      e.preventDefault()
    },

    /**
     * Play the next song when user presses J.
     *
     * @param {Object} e The keydown event
     */
    playNext: e => {
      if ($.is(e.target, 'input, select, textarea')) {
        return true
      }

      playback.playNext()
      e.preventDefault()
    },

    /**
     * Put focus into the search field when user presses F.
     *
     * @param {Object} e The keydown event
     */
    search: e => {
      if (($.is(e.target, 'input, select, textarea') && !$.is(e.target, 'input[type=range]')) || e.metaKey || e.ctrlKey) {
        return true
      }

      e.preventDefault()
      event.emit(event.$names.FOCUS_SEARCH_FIELD)
    },

    /**
     * Like/unlike the current song when use presses L.
     */
    toggleLike: e => {
      if ($.is(e.target, 'input, select, textarea')) {
        return true
      }

      if (!queueStore.current) {
        return
      }

      favoriteStore.toggleOne(queueStore.current)
      socket.broadcast(event.$names.SOCKET_SONG, songStore.generateDataToBroadcast(queueStore.current))
    }
  },

  created: () => KOEL_ENV === 'app' && listenToGlobalShortcuts()
}
</script>
