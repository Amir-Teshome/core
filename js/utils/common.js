/**
 * Other common methods.
 */
import select from 'select'
import { event, use } from '@/utils'
import { sharedStore } from '@/stores'

/**
 * Load (display) a main panel (view).
 *
 * @param {String} view   The view, which can be found under components/main-wrapper/main-content.
 * @param {...*}      Extra data to attach to the view.
 */
export const loadMainView = (view, ...args) => event.emit(event.$names.LOAD_MAIN_CONTENT, view, ...args)

/**
 * Force reloading window regardless of "Confirm before reload" setting.
 * This is handy for certain cases, for example Last.fm connect/disconnect.
 */
export const forceReloadWindow = () => {
  if (window.__UNIT_TESTING__) {
    return
  }

  window.onbeforeunload = () => {}
  window.location.reload()
}

/**
 * Show the overlay.
 *
 * @param  {String}  message
 * @param  {String}  type
 * @param  {Boolean} dismissable
 */
export const showOverlay = (message = 'Just a little patienceâ€¦', type = 'loading', dismissable = false) =>
  event.emit(event.$names.SHOW_OVERLAY, { message, type, dismissable })

/**
 * Hide the overlay.
 */
export const hideOverlay = () => event.emit(event.$names.HIDE_OVERLAY)

/**
 * Copy a text into clipboard.
 *
 * @param  {string} txt
 */
export const copyText = text => use(document.querySelector('#copyArea'), copyArea => {
  copyArea.style.top = `${window.pageYOffset || document.documentElement.scrollTop}px`
  copyArea.value = text
  select(copyArea)
  document.execCommand('copy')
})

export const getDefaultCover = () => `${sharedStore.state.cdnUrl}/public/img/covers/unknown-album.png`

/**
 * Create a fancy ghost drag image.
 *
 * @param {DragEvent} event
 * @param {string} text
 */
export const createGhostDragImage = (event, text) => use(document.querySelector('#dragGhost'), ghost => {
  ghost.innerText = text
  event.dataTransfer.setDragImage(ghost, 0, 0)
})
