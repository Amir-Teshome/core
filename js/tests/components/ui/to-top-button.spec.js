import Component from '@/components/ui/to-top-button.vue'
import { $ } from '@/utils'
import { mock } from '@/tests/__helpers__'

describe('components/ui/to-top-button', () => {
  afterEach(() => {
    jest.resetModules()
    jest.clearAllMocks()
  })

  it('scrolls to top', () => {
    const m = mock($, 'scrollTo')
    shallow(Component).click('button')
    expect(m).toHaveBeenCalled()
  })
})
