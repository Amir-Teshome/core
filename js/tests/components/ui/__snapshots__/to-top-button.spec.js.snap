// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ui/to-top-button renders properly 1`] = `
<transition-stub name="fade">
  <div class="to-top-btn-wrapper" style="display: none;"><button title="Scroll to top"><i class="fa fa-arrow-circle-up"></i> Top
    </button></div>
</transition-stub>
`;
